### PR TITLE
Add some bifunctors-style newtypes

### PR DESCRIPTION
**Pull request by garyb**
The name `Cowrap` is probably a bit odd, but I haven't come up with anything better yet.

### CODE REVIEW

File: src/Data/Profunctor/Join.purs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import Prelude
+
+import Data.Functor.Invariant (class Invariant)
+import Data.Newtype (class Newtype)
+import Data.Profunctor (class Profunctor, dimap)
+
+-- | Turns a `Profunctor` into a `Invariant` functor by equating the two type
+-- | arguments.
+newtype Join p a = Join (p a a)
```
Comment (paf31) [inline src/Data/Profunctor/Join.purs]: I wonder if it's worth providing the adjoint, which constructs a profunctor from an invariant.

Comment (paf31) [reply src/Data/Profunctor/Join.purs]: Personally, I'd rather just remove `Invariant`, but that's a separate, larger conversation 😄

**Comment by paf31** [comment]
👍 LGTM. Curious though - is there a use case for these?

File: src/Data/Profunctor/Join.purs (hunk 1 above)
Comment (garyb) [reply src/Data/Profunctor/Join.purs]: Wait! I just realised something, this is https://github.com/purescript/purescript-monoid/pull/34 if we provide

``` purescript
instance semigroupJoin :: Semigroupoid p => Semigroup (Join p a) where
  append (Join a) (Join b) = Join (a <<< b)

instance monoidJoin :: Category p => Monoid (Join p a) where
  mempty = Join id
```

Comment (paf31) [reply src/Data/Profunctor/Join.purs]: 👍

**Comment by garyb** [comment]
> Curious though - is there a use case for these?

Uhh... most of them are pretty speculative :wink:

It came from a discussion with @safareli, he was asking if we could impose `Functor` and `Contravariant` superclasses, or provide instances for `Functor` and `Contravariant` for every `Profunctor` - since we have endofunctors, no, but I figured providing `newtype`s for them would "prove" the relationship is there, and perhaps be occasionally useful.

Then when I started on that, I figured that profunctors are still a bifunctors, just not in the purescript sense, so I went back to BF to see which of the various newtypes may translate to PF. So `Clown` and `Joker` are definitely a bit 🤷‍♂️, and `Join` was right up until two minutes ago when I realised it's our abstracted `Endo` 😆

File: src/Data/Profunctor/Join.purs (hunk 1 above)
Comment (garyb) [reply src/Data/Profunctor/Join.purs]: Any idea what the adjoint would look like? I just took a stab at it, but not really sure what I'm doing there and couldn't figure out how to implement `dimap`.

**Comment by paf31** [comment]
Do you think the names `Clown` and `Joker` might be confusing though?

**Comment by safareli** [comment]
there is already Joker and Clown defined in https://pursuit.purescript.org/packages/purescript-bifunctors

**Comment by garyb** [comment]
I'm happy to change them, just figured since they work the same way for bifunctors, reusing the names here was reasonable.

File: src/Data/Profunctor/Join.purs (hunk 1 above)
Comment (paf31) [reply src/Data/Profunctor/Join.purs]: Yeah, you can derive something Yoneda-esque if you try to find an adjoint to `Join`:

```purescript
data PronedaF i a b x = PronedaF (a -> x) (i x) (x -> b)
type Proneda i a b = Exists (PronedaF i a b)

instance profunctorProneda :: Profunctor (Proneda i) where ...

lowerProneda :: forall i a b. Proneda i a a -> i a
```

Comment (paf31) [reply src/Data/Profunctor/Join.purs]: I found the previous discussion of this [here](https://github.com/purescript/purescript-monoid/pull/13#issuecomment-79726322).

Comment (garyb) [reply src/Data/Profunctor/Join.purs]: Ah yeah! I remember now. Will add that too then.

Comment (paf31) [reply src/Data/Profunctor/Join.purs]: Feel free to give it a better name 😄

Comment (garyb) [reply src/Data/Profunctor/Join.purs]: `Cojoin` was all I had, I'm not sure it's any better 😆

Comment (paf31) [reply src/Data/Profunctor/Join.purs]: `Split` possibly? Since we're trying to split the type argument into two.

**Comment by garyb** [comment]
There we go, `Split` sounded good to me.

File: src/Data/Profunctor/Split.purs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import Prelude
+
+import Data.Exists (Exists, mkExists, runExists)
+import Data.Functor.Invariant (class Invariant, imap)
+import Data.Profunctor (class Profunctor)
+
+newtype Split f a b = Split (Exists (SplitF f a b))
+
+data SplitF f a b x = SplitF (a -> x) (x -> b) (f x)
```
Comment (paf31) [inline src/Data/Profunctor/Split.purs:18]: Can we hide the constructors here?

**Comment by paf31** [comment]
👍 LGTM, thanks!

File: src/Data/Profunctor/Split.purs (hunk 2 above)
Comment (garyb) [reply src/Data/Profunctor/Split.purs:18]: Oops, good call.

**Comment by paf31** [comment]
I just thought - should this have reused the existing newtypes? Bit late now, but probably worth considering.

**Comment by garyb** [comment]
I just explored that a bit...

- We could reuse `Clown` and `Joker`, defining the instances in `Data.Profunctor`
- `Wrap` doesn't work as it has an overlapping `Functor` instance (one with `Bifunctor` and one with `Profunctor` constraints on `p`)
- `Join` could work, but we'd have to define those "endo" instances with the `newtype`, which would mean some dependency reordering - it probably works out, but `bifunctors` and `monoid` are both near the top of the hierarchy, so I'm a little hesitant to mess with them right now.

I think next time we do breaking releases for all the core libraries we can take a look at it again though :+1:

**Pull request merged**